### PR TITLE
[FIX] Bootnode re-discovery issue

### LINKED ISSUE #695: Unable to execute bootnode peer discovery
### Description

Sometimes my bootnodes reboot and the validators lose connection with the bootnodes

### Your environment

* 0.4.1_linux_amd64

### Steps to reproduce

Run a network with 2 bootnodes and 4 validators.
After stopping all bootnodes and starting the bootnodes again, the validators cannot connect to the bootnodes

### Expected behaviour

Expected would be that the validators would find the bootnodes again

### Actual behaviour

The validators issue an error that they cannot fetch the validators

### Logs

```
2022-08-24T20:11:11.251Z [ERROR] polygon.network.discovery: Unable to execute bootnode peer discovery, %w: EXTRA_VALUE_AT_END="unable to create new discovery client connection, peer disconnected before the discovery client was initialized"
```

### Proposed solution

I would like validators to be able to connect to bootnode without me running the command to add the peer.

When I add a peer to a validator, the network starts working again.

```
polygon-edge peers add --grpc-address 127.0.0.1:10000 --addr
```

### PR DESCRIPTION
**Pull request by ZeljkoBenovic**
# Description

Right now, in a running network, once the bootnodes lose all of its peers (server restart, network issues or similar) it does not get back any peers, without restarting all the nodes.    
There is a go routine that is supposed to ping random bootnode for the list of its peers, but the bootnodes stay disconnected from the rest of the network.

One can test this very easily, run a 4-node cluster with 1 bootnode. Keep all the nodes running but restart the bootnode. Once the bootnode goes back online, it will never join the network and it will always have 0 peers.

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Run a 4-node cluster with 1 bootnode. Restart or power cycle the bootnode only.   
After the bootnode comes back online, after a while (no more than 60s), it should start peering with all the other nodes.

# Additional comments

Fixes EDGE-790

### CODE REVIEW

**Review by epikichi — APPROVED** [review]
Great catch. Confirmed locally that peers will dial back after bootnode(s) disruption occurs.

**Review by Kourin1996 — APPROVED** [review]
LGTM. Thank you for correcting minor things in addition to the fixes

**Review by 0xAleksaOpacic — APPROVED** [review]
LGTM, tested manually everything and the issue is fixed with this

File: network/testing/testing.go [hunk 1]
```diff
@@ -86,6 +86,8 @@ type getRandomPeerDelegate func() *peer.ID
 type fetchAndSetTemporaryDialDelegate func(peer.ID, bool) bool
 type removeTemporaryDialDelegate func(peer.ID)
 
+func (m *MockNetworkingServer) TemporaryDialPeer(peerAddrInfo *peer.AddrInfo) {}
```
Comment (zivkovicmilos) [inline network/testing/testing.go]: Nitpick:

Can you please introduce a delegate method like we did for other `MockNetworkingServer` methods, and populate the body of this callback, like we did for other methods?

It makes it easier in the future to mock, and is consistent with the way we write mocks

File: network/server.go [hunk 2]
```diff
@@ -329,10 +330,17 @@ func (s *Server) checkPeerConnections() {
...
+				randPeer := s.GetRandomPeer()
+				if randPeer != nil && !s.IsConnected(*randPeer) {
+					s.addToDialQueue(s.GetPeerInfo(*randPeer), common.PriorityRandomDial)
+				}
 			} else {
+				// dial random unconnected bootnode
 				randomNode := s.GetRandomBootnode()
-				s.addToDialQueue(randomNode, common.PriorityRandomDial)
+				if randomNode != nil {
+					s.addToDialQueue(randomNode, common.PriorityRandomDial)
```
Comment (zivkovicmilos) [inline network/server.go]: You can inline this

Comment (zivkovicmilos) [reply network/server.go]:
```go
				if randomNode := s.GetRandomBootnode(); randomNode != nil {
					s.addToDialQueue(randomNode, common.PriorityRandomDial)
				}
```

**Review by zivkovicmilos — APPROVED** [review]
Looks good 💯 

Thank you for hunting down this bug 🙏 

I've left a few small comments that need to be addressed.

Please also resolve the conflicts in the `go.mod` file 🙏

File: network/server.go (hunk 2 above)
Comment (ZeljkoBenovic) [reply network/server.go]: Good catch 🎖️  Fixed in: 1531d1aa1f19ed65d2448230e0b026c1488cac2b

File: network/testing/testing.go (hunk 1 above)
Comment (ZeljkoBenovic) [reply network/testing/testing.go]: Fixed in 1531d1aa1f19ed65d2448230e0b026c1488cac2b

**Pull request merged**